lib/atoi/strtou_noneg.h

```diff
@@ -0,0 +1,47 @@
+// SPDX-FileCopyrightText: 2023, Alejandro Colomar <alx@kernel.org>
+// SPDX-License-Identifier: BSD-3-Clause
+
+
+#ifndef SHADOW_INCLUDE_LIB_ATOI_STRTOU_NONEG_H_
+#define SHADOW_INCLUDE_LIB_ATOI_STRTOU_NONEG_H_
+
+
+#include <config.h>
+
+#include <errno.h>
+#include <stdlib.h>
+
+#include "attr.h"
+
+
+ATTR_STRING(1)
+inline unsigned long strtoul_noneg(const char *nptr,
+    char **restrict endptr, int base);
+ATTR_STRING(1)
+inline unsigned long long strtoull_noneg(const char *nptr,
+    char **restrict endptr, int base);
+
+
+inline unsigned long
+strtoul_noneg(const char *nptr, char **restrict endptr, int base)
+{
+	if (strtol(nptr, endptr, base) < 0) {
+		errno = ERANGE;
+		return 0;
+	}
+	return strtoul(nptr, endptr, base);
+}
+
+
+inline unsigned long long
+strtoull_noneg(const char *nptr, char **restrict endptr, int base)
+{
+	if (strtol(nptr, endptr, base) < 0) {
```

Comment:
should this be using strtoll ?

Reply:
Not really.  strtol(3) is good enough there.  Any negative number will be understood by both strtol(3) and strtoll(3) as something `<0`, so both will work.

Reply:
endptr is set to the first non-numeric character in both strtol(3) and strtoll(3), no matter how many digits they find.

Underflow is reported with either LONG_MIN or LLONG_MIN, but both happen to be <0, of course, so it doesn't really matter.  The user of strtoull_noneg() shouldn't be able to notice if it is implemented with strtol(3) or strtoll(3) for rejecting negative values.